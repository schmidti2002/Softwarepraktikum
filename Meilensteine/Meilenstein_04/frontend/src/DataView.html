<style>
    .dataview-type-object {
        border: black 1px solid;
        padding: 4px;
        margin-left: 4px;
    }

    .dataview-type-number {
        color: green;
    }
    
    .dataview-type-string {
        color: orange;
    }
    
    .dataview-type-undefined {
        color: red;
    }

    /*todo other types*/
</style>
<div id="dataview-container" class="row">
    <div id="objContainer" class="col row"></div>
    <div id="simpleDataContainer" class="col"></div>
</div>