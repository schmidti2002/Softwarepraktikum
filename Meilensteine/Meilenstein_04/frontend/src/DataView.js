import View from './View';

// Diese Klasse alle derzeit gespeichtern Variablen und ihren Wert anzeigen
export default class DataView extends View {
  #simpleDataContainer;
  #objContainer;
  #outputDataContainer
  constructor(parentNode, eventReporter) {
    super('DataView', parentNode, eventReporter);
    this.initPromise
      .then(() => {
        this.#objContainer = document.getElementById('objContainer');
        if (!this.#objContainer) {
          eventReporter.fatal('element with id "objContainer" not found!');
        }
        this.#simpleDataContainer = document.getElementById('simpleDataContainer');
        if (!this.#simpleDataContainer) {
          eventReporter.fatal('element with id "simpleDataContainer" not found!');
        }
        this.#outputDataContainer = document.getElementById('outputDataContainer');
        if (!this.#outputDataContainer) {
          eventReporter.fatal('element with id "outputDataContainer" not found!');
        }
      });
  }

  // zeigt die Daten an
  // (unter anderem?) aufgerufen in #onLogicStateChange(data, variables, line, running) in AuDView.js
  renderData(data) {
    if (data === null || data === undefined) {
      this.showEmpty();
      return;
    }
    const divObj = document.createElement('div');
    divObj.classList.add('border');
    const divSimple = document.createElement('div');    
    divObj.classList.add('border');
    const divOutput = document.createElement('div');
    this.#renderData(divObj, divSimple, divOutput, data);
    this.#objContainer.innerHTML = '';
    this.#objContainer.appendChild(divObj);
    this.#simpleDataContainer.innerHTML = '';
    this.#simpleDataContainer.appendChild(divSimple);
    this.#outputDataContainer.innerHTML = '';
    this.#outputDataContainer.appendChild(divOutput);
  }

  #renderData(parentNodeObj, parentNodeSimple, parentNodeOutput, data) {
    const divObj = document.createElement('div');
    divObj.classList.add('row');
    const divSimple = document.createElement('div');
    divObj.classList.add('row');
    const divOutput = document.createElement('div');    
    if(data.output !== null){
      const innerDiv = document.createElement('div');
      innerDiv.classList.add('d-flex');
      innerDiv.classList.add('col');
      innerDiv.classList.add('border');
      innerDiv.textContent = 'output: ';
      this.#renderDataObject(innerDiv, data.output);
      divOutput.appendChild(innerDiv);
    }
    switch (typeof data) {
      case 'object':
        if (!data) {
          divSimple.innerText = String(data);
        }
        Object.keys(data).forEach((key) => {
          if (data[key] !== undefined) {
            const innerDiv = document.createElement('div');
            innerDiv.classList.add('d-flex');
            innerDiv.textContent = `${key}:`;
            // primÃ¤rer Datentyp oder Objekt?
            if (typeof data[key] === 'number' || typeof data[key] === 'string' || typeof data[key] === 'boolean') {
              this.#renderDataSimple(innerDiv, data[key]);
              divSimple.appendChild(innerDiv);
            } else {
              innerDiv.classList.add('col');
              innerDiv.classList.add('border');
              this.#renderDataObject(innerDiv, data[key]);
              divObj.appendChild(innerDiv);
            }
          }
        });
        break;
      default:
        divSimple.textContent = String(data);
    }
    divObj.classList.add(`dataview-type-${typeof data}`);
    parentNodeObj.appendChild(divObj);
    divSimple.classList.add(`dataview-type-${typeof data}`);
    parentNodeSimple.appendChild(divSimple);
    divOutput.classList.add(`dataview-type-${typeof data}`);
    parentNodeOutput.appendChild(divOutput);
  }

  #renderDataSimple(parentNode, data) {
    const div = document.createElement('div');
    div.textContent = String(data);
    div.classList.add(`dataview-type-${typeof data}`);
    parentNode.appendChild(div);
  }

  #renderDataObject(parentNode, data) {
    const div = document.createElement('div');
    switch (typeof data) {
      case 'object':
        if (!data) {
          div.innerText = String(data);
          break;
        }
        Object.keys(data).forEach((key) => {
          if (data[key] !== undefined) {
            const innerDiv = document.createElement('div');
            innerDiv.classList.add('d-flex');
            innerDiv.textContent = `${key}:`;
            this.#renderDataObject(innerDiv, data[key]);
            div.appendChild(innerDiv);
          }
        });
        break;
      default:
        div.textContent = String(data);
    }
    div.classList.add(`dataview-type-${typeof data}`);
    parentNode.appendChild(div);
  }

  /*#renderOutput(parentNode, data){
    const div = document.createElement('div');
    switch (typeof data) {
      case 'object':
        if (!data) {
          div.innerText = String(data);
          break;
        }
        Object.keys(data).forEach((key) => {
          if (data[key] !== undefined) {
            const innerDiv = document.createElement('div');
            innerDiv.classList.add('d-flex');
            innerDiv.textContent = `${key}:`;
            this.#renderDataObject(innerDiv, data[key]);
            div.appendChild(innerDiv);
          }
        });
        break;
      default:
        div.textContent = String(data);
    }
    div.classList.add(`dataview-type-${typeof data}`);
    parentNode.appendChild(div);
  }*/

  showEmpty() {
    this.#objContainer.innerHTML = 'Placeholder for nothing to show in objContainer';
    this.#simpleDataContainer.innerHTML = 'Placeholder for nothing to show in simpleDataContainer';
    this.#outputDataContainer.innerHTML = 'Placeholder for nothing to show in outputDataContainer';
  }
}
