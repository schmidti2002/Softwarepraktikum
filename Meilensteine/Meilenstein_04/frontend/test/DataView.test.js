import { describe, test, expect, jest, beforeEach, afterEach } from '@jest/globals';
import DataView from '../src/DataView';
import { mockFetchHtml } from './testUtils.test';

describe('DataView', () => {
  let dataView;
  let mockContainer;
  let mockEventReporter;

  beforeEach(async () => {
    mockFetchHtml('../src/DataView.html');
    
    mockEventReporter = {
      fatal: jest.fn(),
    };

    dataView = new DataView(document.body, mockEventReporter);
    mockContainer = document.getElementById('dataview-container');
    await dataView.initPromise; 

    jest.spyOn(document, 'getElementById');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe('constructor', () => {
    test('initializes with provided parentNode and mockEventReporter', async () => {
      expect(dataView).toBeDefined();
      expect(mockContainer).toBeDefined(); 
    });
  });

  describe('error handling', () => {
    test('handles missing container element gracefully', async () => {
      document.getElementById.mockReturnValueOnce(null);
      const localDataView = new DataView(document.body, mockEventReporter);
      await localDataView.initPromise;
      expect(mockEventReporter.fatal).toHaveBeenCalled(); 
    });
  });

  describe('showEmpty method', () => {
    test('displays "Placeholder for nothing to show"', async () => {
      dataView.showEmpty();
      expect(mockContainer.innerHTML).toBe('Placeholder for nothing to show');
    });
  });

  describe('renderData method', () => {
    test.each([
      [{ key: 'value' }, 'key:', 'value', '.dataview-type-object'],
      [123, '123', null, '.dataview-type-number'],
      ['test string', 'test string', null, '.dataview-type-string'],
      [[], '', null, null],
      [{}, '', null, null],
      [null, 'Placeholder for nothing to show', null, null],
      [undefined, 'Placeholder for nothing to show', null, null],
      [['one', 'two', 'three'], 'one', 'two', null],
      [{ parent: { child: 'value' } }, 'parent:', 'child:', '.dataview-type-object'],
    ])('renders %p correctly', async (input, expectedInnerHTML, expectedValue, expectedClass) => {
      dataView.renderData(input);
      expect(mockContainer.innerHTML).toContain(expectedInnerHTML);
      if (expectedValue !== null) {
        expect(mockContainer.innerHTML).toContain(expectedValue);
      }
      if (expectedClass !== null) {
        const renderedElements = mockContainer.querySelectorAll(expectedClass);
        expect(renderedElements.length).toBeGreaterThan(0);
      }
    });
  });

  describe('handling of null and undefined data', () => {
    test('calls showEmpty for null data', async () => {
      dataView.showEmpty = jest.fn();
      dataView.renderData(null);
      expect(dataView.showEmpty).toHaveBeenCalled();
    });

    test('calls showEmpty for undefined data', async () => {
      dataView.showEmpty = jest.fn();
      dataView.renderData(undefined);
      expect(dataView.showEmpty).toHaveBeenCalled();
    });
  });

  describe('Styling of rendered data based on data type', () => {
    const testDataTypes = [
      { type: 'object', data: { key: 'value' }, expectedClass: '.dataview-type-object' },
      { type: 'number', data: 123, expectedClass: '.dataview-type-number' },
      { type: 'string', data: "Hello", expectedClass: '.dataview-type-string' },
    ];
  
    testDataTypes.forEach(({ type, data, expectedClass }) => {
      test(`applies correct class for ${type} type`, async () => {
        dataView.renderData(data);
        const renderedElements = mockContainer.querySelectorAll(expectedClass);
        expect(renderedElements.length).toBeGreaterThan(0);
      });
    });
  })

  describe('Snapshot Tests', () => {
    test('snapshot with simple data object', () => {
      const testData = { key: 'value' };
      dataView.renderData(testData);
      expect(mockContainer.innerHTML).toMatchSnapshot();
    });

    test('snapshot with nested data object', () => {
      const testData = { parent: { child: 'value' } };
      dataView.renderData(testData);
      expect(mockContainer.innerHTML).toMatchSnapshot();
    });

    test('snapshot with array data', () => {
      const testData = ['one', 'two', 'three'];
      dataView.renderData(testData);
      expect(mockContainer.innerHTML).toMatchSnapshot();
    });
  });
});