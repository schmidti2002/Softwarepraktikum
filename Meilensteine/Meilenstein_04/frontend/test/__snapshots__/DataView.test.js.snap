// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataView Snapshot Tests snapshot with array data 1`] = `"<div><div class="dataview-type-object"><div class="d-flex"><div class="dataview-type-object"><div class="d-flex"><div class="dataview-type-string"></div></div></div></div></div></div>"`;

exports[`DataView Snapshot Tests snapshot with nested data object 1`] = `"<div><div class="dataview-type-object"><div class="d-flex"><div class="dataview-type-string"></div></div></div></div>"`;

exports[`DataView Snapshot Tests snapshot with simple data object 1`] = `"<div><div class="dataview-type-string"></div></div>"`;
